File: flow/src/main/java/flow/HistoryCallback.java
package flow;

import android.app.Activity;
import androidx.annotation.NonNull;

/**
 * An object to which gets a chance to modify the current {@link History} before
 * it is persisted, or after it is restored.
 */
public interface HistoryCallback {
  @NonNull
  default History onRestoreHistory(@NonNull History history) {
    return history;
  }

  @NonNull
  default History onSaveHistory(@NonNull History history) {
    return history;
  }

  /**
   * Called when history is cleared.
   * Default implementation calls {@link Activity#finish()}}
   */
  default void onHistoryCleared() {
  }
}

File: flow/src/main/java/flow/HistoryCallback.kt
package flow

import android.app.Activity

/**
 * An object to which gets a chance to modify the current [History] before
 * it is persisted, or after it is restored.
 */
interface HistoryCallback {
  fun onRestoreHistory(history: History): History =
    history

  fun onSaveHistory(history: History): History =
    history

  /**
   * Called when history is cleared.
   * Default implementation calls [Activity.finish]
   */
  fun onHistoryCleared() = Unit
}